src/Altinn.Platform/Altinn.Platform.Storage/Storage/Controllers/InstancesController.cs

```diff
@@ -59,33 +66,154 @@ public async Task<ActionResult> GetInstanceOwners(int instanceOwnerId)
                 {
-                    return NotFound($"Did not find any instances for application owner (org)={org}");
+                    return NotFound($"Did not find any instances");
                 }
 
-                return Ok(result);
-            }
-            else if (!string.IsNullOrEmpty(appId))
-            {                
-                List<Instance> result = await _instanceRepository.GetInstancesOfApplication(appId);
-                if (result == null || result.Count == 0)
+                if (!string.IsNullOrEmpty(result.Exception))
                 {
-                    return NotFound($"Did not find any instances for applicationId={appId}");
+                    return BadRequest(result.Exception);
                 }
+          
+                string nextContinuationToken = HttpUtility.UrlEncode(result.ContinuationToken);
+                result.ContinuationToken = nextContinuationToken;
+                result.ContinuationToken = null;
 
-                return Ok(result);
+                HALResponse response = new HALResponse(result);
+
+                if (continuationToken == null)
+                {
+                    string selfUrl = $"{host}{url}{query}";
+
+                    result.Self = selfUrl;
+
+                    Link selfLink = new Link("self", selfUrl);
+                    response.AddLinks(selfLink);
+                }
+                else
+                {
+                    string selfQueryString = BuildQueryStringWithOneReplacedParameter(
+                        queryParams,
+                        "continuationToken",
+                        selfContinuationToken);
+
+                    string selfUrl = $"{host}{url}{selfQueryString}";
+
+                    result.Self = selfUrl;
+
+                    Link selfLink = new Link("self", selfUrl);
+                    response.AddLinks(selfLink);
+                }
+
+                if (nextContinuationToken != null)
+                {
+                    string nextQueryString = BuildQueryStringWithOneReplacedParameter(
+                        queryParams,
+                        "continuationToken",
+                        nextContinuationToken);
+
+                    string nextUrl = $"{host}{url}{nextQueryString}";
+
+                    result.Next = nextUrl;
+
+                    Link nextLink = new Link("next", nextUrl);
+                    response.AddLinks(nextLink);
+                }
+
+                // add self links to platform
+                result.Instances.ForEach(i =>
+                {
+                    i.SelfLinks = new ResourceLinks
+                    {
+                        Platform = $"{host}{url}/{i.Id}"
+                    };
+                });
+
+                StringValues acceptHeader = Request.Headers["Accept"];
+                if (acceptHeader.Any() && acceptHeader.Contains("application/hal+json"))
+                {
+                    // convert to HAL
```

Comment:
a comment to remember to implement or to explain what is done here?

Reply:
if it's for explaining the code it might benefit for a couple extra word. What are we converting? the result?

Reply:
Have added comment